### ISSUE TITLE
Using wrong GPU adapter

### ISSUE THREAD

**Issue by kimlar** [grimfang4/sdl-gpu#17]
Hi!
My laptop has one integrated Intel HD 4600 GPU and one NVidia GTX850M. It is using the Intel GPU which is slow (but saves power), but I would like the user to be able to choose which GPU to use.
I just wanted to inform about this issue.

Luckily I found a solution. With this code the NVidia driver chooses the most powerful GPU:

`#include <windows.h> // <---- for the DWORD`
` `
`// enable optimus!`
`extern "C" {`
`	_declspec(dllexport) DWORD NvOptimusEnablement = 0x00000001;`
`}`

I did some bench marking on my laptop: (screen: 800x600, image: 64x64)
For Intel HD: ~10000 blits/sec @ 60 FPS
For NVidia: ~50000 blits/sec @ 60 FPS (Temp: GPU:~75 Celsius, CPU: ~80 Celsius)

I wish there could have been a way to list all adapters and choose which one to use. This would be a good thing for having in the option menu to select wanted adapter.

**Comment by grimfang4** [comment]
Typically, the user would choose by launching the app via right-click > Run with graphics processor or by setting the program up as using the discrete GPU in the Nvidia settings panel.

I hadn't seen any code to do it before, so this is interesting.  I might be able to add equivalent code for AMD and make a new function for this.

**Comment by grimfang4** [comment]
I'm not sure when the driver checks this variable and I no longer have an Optimus laptop.  Can you test something like this?  Ideally, we'd be able to choose before init.  Would you please see if this can control enabling and disabling if used before GPU_Init()?

```
extern "C" {
    _declspec(dllexport) DWORD NvOptimusEnablement = 0x00000000;
    _declspec(dllexport) int AmdPowerXpressRequestHighPerformance = 0;
}

void EnableHighPerformanceGPU(int enable)
{
    AmdPowerXpressRequestHighPerformance = NvOptimusEnablement  = (enable? 1 : 0);
}
```

**Comment by kimlar** [comment]
I did test it and it did not work because it enabled Optimus what ever the
value was. It looks like it is activated as soon as the _declspec(dllexport)
DWORD NvOptimusEnablement was added in the code, doesn't even need an
assigned value.

I found this information:
https://devtalk.nvidia.com/default/topic/812387/nvoptimusenablement-value-not-considered/

But unfortunately does still activate optimus when set to 0x00000000. I am
not sure if NVidia has a bug in their drivers, but according to the
document it should ignore optimus if set to 0x00000000.

Only solution I could think of is to have an external DLL just for the
optimus, if user wants it active. It is kinda an extreme method but I will
see if I could do this.


2016-04-18 22:25 GMT+02:00 Jonathan Dearborn <notifications@github.com>:
